### PR TITLE
HPCC-12883 Remove socket fd max from select and epoll

### PR DESCRIPTION
**Pull request by mckellyln**
This PR uses poll() in place of select() (except on Windows where there is a known won't fix issue with WSAPoll()) AND this PR also uses an array of pointers into heap storage for the SelectItem list of descriptors - so there is no longer indexing by fd into an array (epfdtbl) and thus no longer any limit on the number or value of socket descriptors.  The change for using an array of pointers is referenced in JIRA issue HPCC-10235.  Both issues are fixed here because both relate to a max fd value.

@ghalliday and/or @jakesmith please review.

Signed-off-by: Mark Kelly mark.kelly@lexisnexis.com

### CODE REVIEW

File: system/jlib/jsocket.cpp [hunk 1]
```diff
@@ -3744,29 +3772,42 @@ class CSocketBaseThread: public Thread
...
+            return false;
+        }
+        else if (rc>0)
+        {
+            if (fds[0].revents & POLLNVAL)
+            {
+                StringBuffer sockstr;
+                const char *tracename = sockstr.append((unsigned)sock).str();
+                LOGERR2(ERRNO(),3,"CSocketBaseThread poll handle");
+                return false;
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: trivial: would be nice if the above 2 identical block were common up.

File: system/jlib/jsocket.cpp [hunk 2]
```diff
@@ -4316,19 +4396,37 @@ class CSocketSelectHandler: public CInterface, implements ISocketSelectHandler
...
+
+        int fd = si->handle;
+
+        if (fd < 0)
+        {
+# ifdef EPOLLTRACE
+            DBGLOG("EPOLL: op(%d) bad fd %d to epfd %d", op, fd, efd);
+# endif
+            return;
+        }
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: Are the above conditions possible?

File: system/jlib/jsocket.cpp [hunk 3]
```diff
@@ -3965,6 +4006,45 @@ class CSocketSelectThread: public CSocketBaseThread
         basesize = n;
     }
 
+    bool checkSocks()
+    {
+        bool ret = false;
+        ForEachItemIn(i,items) {
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: formatting: block on newline

(A few more below)

File: system/jlib/jsocket.cpp [hunk 4]
```diff
@@ -4415,7 +4514,8 @@ class CSocketEpollThread: public CSocketBaseThread
         validateerrcount = 0;
         offset = 0;
         selecttrace = trc;
-        epfd = ::epoll_create(XFD_SETSIZE);
+        max_ret_events = 5000;
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: minor: perhaps should be a constant defined elsewhere (e.g. at head of file), with a comment describing significance.

File: system/jlib/jsocket.cpp [hunk 5]
```diff
@@ -4447,14 +4537,19 @@ class CSocketEpollThread: public CSocketBaseThread
...
-                si.sock->Release();
+        ForEachItemIn(i,items)
+        {
+            try
+            {
+                SelectItem *si = items.element(i);
+                epoll_op(epfd, EPOLL_CTL_DEL, si, 0);
+                si->nfy->Release();
+                si->sock->Release();
+                delete si;
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: It might be cleaner here and in other places, if a ~SelectItem handled releasing it's linked items (nfy, sock)

Comment (jakesmith) [inline system/jlib/jsocket.cpp]: It might also be worth making SelectItem a owned object and have the items array own them for tidier clearup.

File: system/jlib/jsocket.cpp [hunk 6]
```diff
@@ -4476,105 +4570,116 @@ class CSocketEpollThread: public CSocketBaseThread
...
-        SelectItem sn;
-        sn.nfy = LINK(nfy);
-        sn.sock = LINK(sock);
-        sn.mode = (byte)mode;
-        sn.handle = (T_SOCKET)sock->OShandle();
-        CHECKSOCKRANGE(sn.handle);
-        sn.del = false;
-        sn.add_epoll = true;
+        CriticalBlock block(sect);
+        if (sock==NULL) { // wait until no changes outstanding
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: formatting: brace on newline.

**Comment by jakesmith** [comment]
@mckellyln - looks good, minor comments only.

**Comment by mckellyln** [comment]
@jakesmith thanks, addressed most of your comments with additional commit.

File: system/jlib/jsocket.cpp (hunk 5 above)
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: I agree SelectItem dtor could do nfy/sock Release but I was hesitant to change any of the select method code.

File: system/jlib/jsocket.cpp (hunk 3 above)
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: Change in many places, but left select method code as is.

File: system/jlib/jsocket.cpp (hunk 1 above)
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: Done.

File: system/jlib/jsocket.cpp (hunk 3 above)
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: Ok, my policy is that if I touch a method I correct the formatting at that point. That sometimes extends to if I touch a class.

File: system/jlib/jsocket.cpp (hunk 5 above)
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: fair enough. Could be another PR anyway.

File: system/jlib/jsocket.cpp (hunk 2 above)
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: ok, probably not - removed.

**Comment by mckellyln** [comment]
@jakesmith I've removed the extra checks in epoll_op() and I've changed a few more single link block start formatting for the methods in CSocketEpollThread class.

**Comment by jakesmith** [comment]
Looks good, but should this be a 6.2 change not a point change?
Not needed for 6.0.x (afaik) and _may_ bring with it unknown issues potentially.
@richardkchapman

**Comment by richardkchapman** [comment]
Feels like a 6.2 change to me

**Comment by mckellyln** [comment]
closing to retarget 6.2.0 (master)
